Usermode prep/fixes

eim64 opened the pull request:
Fixes tss stack to avoid triple fault
Make gdt compatible with syscall/sysret
Set user bit on all pages with prot != 0 so that memory can be used outside of ring 0

antlilja commented on kernel/src/memory/paging.c:
```diff
@@ -125,6 +125,7 @@ void map_address_space(AddressSpace* space) {
 
     g_pml4[space->pdp_index].present = true;
     g_pml4[space->pdp_index].write = true;
+    g_pml4[space->pdp_index].user = space->prot != 0;
```
These changes to paging.c are already in the main branch

antlilja commented on kernel/src/gdt.c:
```diff
@@ -20,8 +20,8 @@ struct {
     GDTEntry kernel_code;
     GDTEntry kernel_data;
     GDTEntry null2;
-    GDTEntry user_code;
```
Why has the code and data segment switched places?

antlilja commented on kernel/src/gdt.c:
```diff
@@ -97,6 +97,9 @@ void setup_gdt_and_tss() {
     g_tss.rsp[0] = (void*)alloc_pages(TSS_STACK_PAGES, PAGING_WRITABLE);
 
+    // Move stack to end of allocation
+    g_tss.rsp[0] += TSS_STACK_PAGES * PAGE_SIZE;
```
You should probably just do this on the line where the stack is being allocated instead of splitting it in two. 

The IST pointer should also start at the bottom as well.

eim64 committed:
change order of user gdt entries
syscall sets ss = base +0x08 and cs = base + 0x10

therefore, data segment must be before code segment for syscall/sysret

"Upon return, SYSRET copies the value saved in RCX to the RIP.

In a return to 64-bit mode using Osize 64, SYSRET sets the CS selector value to
 MSR_IA32_STAR[63:48] +16. The SS is set to IA32_STAR[63:48] + 8.

eim64 committed:
move stack pointers to end of allocation

stack uses sub, which means any interrupt would
trigger a pagefault interrupt, which triggers a pagefault interrupt.....
